### ISSUE TITLE
[css-values] Clarify fragment URLs resolve against the current tree, not document tree

### ISSUE THREAD

**Issue by rniwa** [w3c/csswg-drafts#3320]
Right now, [section 4.4.1.1. fragment URLs](https://drafts.csswg.org/css-values-4/#local-urls) say that:
> If a url()’s value starts with a U+0023 NUMBER SIGN (#) character, parse it as per normal for URLs, but additionally set the local url flag of the url().
>
> When matching a url() with the local url flag set, ignore everything but the URL’s fragment, and resolve that fragment against the current document that relative URLs are resolved against. This reference must always be treated as same-document (rather than cross-document).

We text should be updated to refer to the current tree so that when the [local url flag](https://drafts.csswg.org/css-values-4/#url-local-url-flag) is set, the fragment is resolved against the current tree.

**Comment by rniwa** [comment]
Also see https://github.com/w3c/csswg-drafts/issues/2715

**Comment by rniwa** [comment]
Also see https://github.com/w3c/csswg-drafts/issues/1995

**Comment by tabatkins** [comment]
All right seems like the webapps people were pretty united on this a few years ago, so I've gone ahead and made this change.

**Comment by fantasai** [comment]
Re-opening for CSSWG resolution.
Note that neither Blink nor Firefox behave as specced, see [testcase](https://www.software.hixie.ch/utilities/js/live-dom-viewer/?%3C!DOCTYPE%20html%3E%0A%3Cbase%20href%3D%22https%3A%2F%2Fxanthir.com%2Fpony%22%3E%0A%3Ca%20href%3D%22%23linky%22%20style%3D%22background%3A%20url(%23foo)%3B%20font-size%3A%202em%22%3Etest%3C%2Fa%3E%0A) which shows that they are currently honoring `<base href=...>` when resolving fragment URLs.

**Comment by tabatkins** [comment]
Yeah, I'm not sure I understand what was being argued here either (I unfortunately had to miss this meeting).

**Comment by emilio** [comment]
@annevk @tabatkins my point was that there are two orthogonal issues:

 * Whether `<base>` is respected or we should just ignore and treat as local refs.
 * Which tree we find stuff in. And it's unclear browsers are consistent, I'm pretty sure we don't use the stylesheet's tree.

I guess you would be opposed to honor `<base>` and lookup inside shadow trees if the document uri matches?

**Comment by tabatkins** [comment]
Your first question was resolved back in the day, when we discussed this original behavior and I added the "local url" flag. Browser support was uncertain at the time but it was considered desirable to make the fragment URLs local-only.

The only new question was whether they were truly fragments resolved against the current page (thus only capable of resolving to IDs visible in the topmost light DOM) or if they were tree-specific (so you could target IDs in your shadow). The Webapps folks seemed reasonably consistent in wanting the latter, and that made sense to me, so I updated the spec here.

Apologies for not being around to introduce the topic properly, this was meant to just be a review of the most recent change for box-ticking purposes, not a relitigation of the earlier discussions.

**Comment by annevk** [comment]
> I guess you would be opposed to honor `<base>` and lookup inside shadow trees if the document uri matches?

Yeah that breaks encapsulation.

**Comment by tabatkins** [comment]
Okay, new text committed as a result of the WG discussion - fragment-only URLs are now explicitly treated as tree-scoped references, with the set of IDs in the tree as the associated tree-scoped names.

This implies that `x-foo::part(foo) { mask: url(#foo); }` will resolve to the element with id=foo in the light DOM, regardless of what IDs are used inside the shadow, while `<div part=foo style="mask: url(#foo)">` inside the shadow will resolve to the element with id=foo in the shadow. (But if there is no such element, it'll walk up to the light and look for the ID there.)

For the question of what to do with Media Fragments and similar non-ID fragment syntaxes, I'm currently just treating them identically to anything else - I try to find an element with a matching ID and if it fails (as it almost certainly will) it fails as normal, and is treated the same as if `#foo` wasn't in the document.

This probably needs some more work to make this good. In particular, this doesn't work properly with [fragment directives](https://wicg.github.io/scroll-to-text-fragment/#processing-the-fragment-directive) - `url(#foo:~:text=foo)` will search for an element with `id="foo:~:text=foo"`, rather than stripping the directive. I've opened an issue (<https://github.com/WICG/scroll-to-text-fragment/issues/198>) to get an algo exposed for doing that easily.

Anything else the WG thinks is still wrong, or needs to be added?

**Comment by fantasai** [comment]
If the intention of this special behavior is to apply this to ID fragments, then scope it to ID fragments. There are other types of fragment identifiers, scroll-to-text-fragment is only one, see e.g. https://www.w3.org/TR/media-frags/ for some others. In general URL fragments are an extension point. Even if we special-case some things, we need to define their generic behavior in a way that's going to work generically.

**Comment by tabatkins** [comment]
> If the intention of this special behavior is to apply this to ID fragments, then scope it to ID fragments. 

I would if I could. The problem is that as far as I can tell, there isn't a well-defined notion of "ID fragments".

**Comment by fantasai** [comment]
Agenda+ to review the latest spec text:
https://drafts.csswg.org/css-values-4/#local-urls
